### ISSUE TITLE
Measure time it took to deliver http.ServerResponse to the client

### ISSUE THREAD

**Issue by pps83** [nodejs/help#734]
I serve large files (10MB+) using nodejs and I would like me measure download speed for each request. However, I do not see anything that would do that for me. In my case I serve dynamically generated content, essentially I send a large nodejs Buffer object. From node's [http.ServerResponse docs](https://nodejs.org/api/http.html#http_event_finish) `finish` is the last emitted event for a server response event:

> After this event, no more events will be emitted on the response object.

In my case, I get 'finish' event after a few milliseconds from receiving incoming request. What can be done in node to measure speed at which remote client downloads data from nodejs http server?

So, far the only thing that works *sometimes* is handling [`drain`  event on the socket](https://nodejs.org/api/net.html#net_event_drain). This is flaky solution, because sometimes for entire connection you don't get even a single `drain` event.

**Comment by tniessen** [comment]
The documentation of the `finish` event explicitely says:

> More specifically, this event is emitted when the last segment of the response headers and body have been handed off to the operating system for transmission over the network. It does not imply that the client has received anything yet.

I am not sure whether our implementation allows to access information beyond that. cc @nodejs/http

**Comment by pps83** [comment]
```
const express = require('express');
const app = express();

app.use(function(req, res, next) {
    if (req.url != '/test')
        return next();
    var t0 = Date.now();
    function log(type) {
        console.log(type + ' time:' + (Date.now() - t0) + ' sock.bytesWritten:' + req.socket.bytesWritten + ' sock.bufferSize:' + req.socket.bufferSize);
    }
    res.setHeader('Content-Length', 25000000);
    res.writeHead(200);
    res.write(Buffer.alloc(25000000, 'a'));
    res.end(() => { log('res.end'); });
    res.on('finish', () => { log('finish'); });
    var int = setInterval(() => { log('timer'); }, 1000);
    res.socket.on('close', () => {
        clearInterval(int);
        log('close');
    });
});
app.listen(3000, function() {
    console.log('Example app listening on port 3000!')
});
```

run it on non-localhost and try to `wget example.com:3000/test`

```
pps@fr:~/projects/file-server$ node test.js
Example app listening on port 3000!
res.end time:41 sock.bytesWritten:25000129 sock.bufferSize:49922620
finish time:42 sock.bytesWritten:25000129 sock.bufferSize:49922620
timer time:1040 sock.bytesWritten:25000129 sock.bufferSize:49922620
timer time:2041 sock.bytesWritten:25000129 sock.bufferSize:49922620
timer time:3042 sock.bytesWritten:25000129 sock.bufferSize:49922620
timer time:4042 sock.bytesWritten:25000129 sock.bufferSize:49922620
timer time:5043 sock.bytesWritten:25000129 sock.bufferSize:49922620
close time:5809 sock.bytesWritten:25000129 sock.bufferSize:undefined
```

Looks like `finish` and end are called immediately (which kind of makes them useless?).
It takes a few more seconds before socket close event is called, which indicates obviously that data was delivered. But.. I don't think that you can pass 25MB to a socket (unless it's a blocking send), there a limits, and I think they are way lower than 25MB. I use ubuntu server and using pure c code I saw that buffers won't grow over 5MB (using blocking sends)

In other words, seems like there is no way to write in node anything network related where 1) it's needed to estimate current transfer speed of a http (tcp) connection, 2) there is no way at all to know if 100bytes or 25MB was sent out over the wire., 3) there is no way to know what what part of it was ACKed by the remote (you may sent out 25MB over the wire that was cut off during transfer). Just like FIONREAD tells how much data is available for read there is also SIOCOUTQ to know how much data is still waiting to be sent out.

**Comment by pps83** [comment]
I'm looking at 6.11.x node's code and this is what I see. Socket is a Duplex stream. Socket.write simply passes that to stream.Duplex.prototype.write, which will buffer data if it's over highWatermark (which is 16K only). Code that writes to a stream should stop writing data to a stream if Duplex.write returns false, and at that point code should wait for 'drain' event that notifies that code should start writing again.
So, on code inspection, I see that return value from Duplex.write is ignored, and I cannot know if my code should continue writing even more data after I wrote 25MB. Also, I don't see at all any hints in the code how to know what data was actually sent to socket and what data is buffered by node. Looks like bytesWritten is simply that total size that was written to net.Socket, not the kernel socket descriptor.
If you attach to `req.socket.on('drain', () => { log('drain'); });` in my code sample you'll see that drain event from Duplex stream notifying that internaly buffer is drained and it's ok to start writing more data. 
So... res.on('finish', ...) must not be called before 'drain' event. And I think it's a bug it node's socket code: it has to wait properly for data to be written to kernel socket descriptor before firing 'finish' event.

**Comment by pps83** [comment]
One more point. [Response.write](https://nodejs.org/api/http.html#http_response_write_chunk_encoding_callback) takes a callback that "will be called when this chunk of data is flushed". Without specifying what it means, I guess it means that this buffer was written to actual socket. So, if I specify these callbacks they will be called way later than I get entire 'finish' event which must must not come before these all these callbacks were called.

I think 1) node's socket/net code has to be fixed to properly emit 'finish' event. 2) there should be another one (or even two events added) that actually notify that real socket's buffer was sent out (and ACKed by remote). On linux SIOCOUTQ could be used, on latest win10 SIO_TCP_INFO can specify TCP_INFO_v0.BytesOut. Basically, it's easy to get kernel socket to buffer 5MB it would be nice to have event when all that data was sent out, all that libuv (or whatever socket async code is used by node) is supposed to do that IMO.

**Comment by gireeshpunathil** [comment]
@pps83 - is this still an issue for you, can you try with the latest to see how it is fairing?

**Comment by gireeshpunathil** [comment]
closing due to inactivity, please re-open if it is still outstanding
